lib/pure/asynchttpserver.nim

```diff
@@ -178,7 +178,12 @@ proc processClient(client: AsyncSocket, address: string,
         except ValueError:
           asyncCheck request.respondError(Http400)
           continue
-      of 1: parseUri(linePart, request.url)
+      of 1: 
+        try:
+          parseUri(linePart, request.url)
+        except ValueError:
+          asyncCheck request.respondError(Http400) 
```

Comment:
Are these `asyncCheck`s really correct here? Should not we block further socket processing with `await`s instead? Also should not we close the connection after errors?

Reply:
The reason for the `asyncCheck` here is that `await` cannot be put inside an `except` branch (IIRC).

Responding with Http400 is cool. The client should be given a chance to retry (as above).

Reply:
> await cannot be put

That is not an excuse for the code that behaves incorrectly. What if both HTTP method and URL are incorrect? Shall we reply twice?

Reply:
I don't think that can happen. The ``continue`` ensures it.

Reply:
It does happen, and results in complete garbage:
```sh
$ telnet localhost 5000
Trying 127.0.0.1...
Connected to localhost.
Escape character is '^]'.
BLA BLA BLA
HTTP/1.1 400 Bad Request
Content-Length: 15

400 Bad RequestHTTP/1.1 400 Bad Request
Content-Length: 15

400 Bad RequestHTTP/1.1 400 Bad Request
Content-Length: 15

400 Bad Request
HTTP/1.1 404 Not Found
Content-Length: 178
Content-Type: text/html;charset=utf-8

<html xmlns="http://www.w3.org/1999/xhtml"><head><title>404 Not Found</title></head><body style="text-align: center;"><h1>404 Not Found</h1><hr/><p>Jester 0.1.0</p></body></html>Connection closed by foreign host.
```

Reply:
This bug is very critical IMO. This could potentially result in clients getting `404` for urls that are actually valid.

Reply:
It's hard to see what you're sending to the server in your session. Can you create an issue for this and elaborate?

Reply:
#6386